browser/default_protocol_handler_utils_win.cc

```diff
@@ -0,0 +1,457 @@
+  const bool is_protocol = protocol[0] != L'.';
+  hr = registration->QueryCurrentDefault(
+      protocol.data(), is_protocol ? AT_URLPROTOCOL : AT_FILEEXTENSION,
+      AL_EFFECTIVE, &current_app);
+  if (FAILED(hr)) {
+    LOG(ERROR) << "Failed to query default app for protocol " << protocol;
+    return std::wstring();
+  }
+  return current_app.get();
+}
+
+// NOTE: The passed-in current user SID is used here, instead of getting the SID
+// for the owner of the key. We are assuming that this key in HKCU is owned by
+// the current user, since we want to replace that key ourselves. If the key is
+// owned by someone else, then this check will fail; this is ok because we would
+// likely not want to replace that other user's key anyway.
+bool CheckUserChoiceHash(base::WStringPiece protocol,
+                         base::WStringPiece user_sid) {
+  auto key_path = GetAssociationKeyPath(protocol.data());
+  if (!key_path)
+    return false;
+
+  base::win::RegKey user_choice_key(HKEY_CURRENT_USER, key_path.get(),
+                                    KEY_READ);
+  if (!user_choice_key.Valid())
+    return false;
+
+  if (user_choice_key.OpenKey(kUserChoiceKey, KEY_READ) != ERROR_SUCCESS)
+    return false;
+
+  FILETIME last_write_file_time;
+
+  // RegKey doesn't support for fetching last write time.
+  if (::RegQueryInfoKeyW(user_choice_key.Handle(), nullptr, nullptr, nullptr,
+                         nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+                         nullptr, &last_write_file_time) != ERROR_SUCCESS) {
+    return false;
+  }
+
+  SYSTEMTIME last_write_system_time;
+  if (!::FileTimeToSystemTime(&last_write_file_time, &last_write_system_time))
+    return false;
+
+  // Read ProgId
+  std::wstring prog_id;
+  if (user_choice_key.ReadValue(kProgIDKey, &prog_id) != ERROR_SUCCESS)
+    return false;
+
+  // Read Hash
+  std::wstring stored_hash;
+  if (user_choice_key.ReadValue(kHashKey, &stored_hash) != ERROR_SUCCESS)
+    return false;
+
+  auto computed_hash = GenerateUserChoiceHash(protocol, user_sid, prog_id,
+                                              last_write_system_time);
+  if (computed_hash.empty())
+    return false;
+
+  if (::CompareStringOrdinal(computed_hash.c_str(), -1, stored_hash.c_str(), -1,
+                             FALSE) != CSTR_EQUAL) {
+    return false;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+std::wstring GenerateUserChoiceHash(base::WStringPiece ext,
+                                    base::WStringPiece sid,
+                                    base::WStringPiece prog_id,
+                                    SYSTEMTIME timestamp) {
+  auto user_choice = FormatUserChoiceString(ext, sid, prog_id, timestamp);
+  if (!user_choice) {
+    LOG(ERROR) << "Didn't get user choice string for generating hash.";
+    return std::wstring();
+  }
+
+  return HashString(user_choice.get());
+}
+
+bool SetDefaultProtocolHandlerFor(base::WStringPiece protocol) {
```

Comment:
both these public APIs should contain `ScopedBlockingCall scoped_blocking_call(FROM_HERE, BlockingType::MAY_BLOCK);` to make sure we use them properly.

Reply:
Fixed.